src/java/org/apache/cassandra/db/compaction/BackgroundCompactionRunner.java

```diff
@@ -0,0 +1,457 @@
+        Collections.shuffle(compactionRequestsList, random);
+
+        for (ColumnFamilyStore cfs : compactionRequestsList)
+        {
+            if (ongoingCompactions.get() >= compactionExecutor.getMaximumPoolSize())
+            {
+                logger.trace("Background compaction threads are busy; delaying new compactions check until there are free threads");
+                return;
+            }
+
+            FutureRequestResult promise = compactionRequests.remove(cfs);
+            assert promise != null : "Background compaction checker must be single-threaded";
+
+            if (promise.isDone())
+            {
+                // A shutdown request may abort processing while we are still processing
+                assert promise.join() == RequestResult.ABORTED : "Background compaction checker must be single-threaded";
+
+                logger.trace("The request for {} was aborted due to shutdown", cfs);
+                continue;
+            }
+
+            if (!cfs.isValid())
+            {
+                logger.trace("Aborting compaction for dropped CF {}", cfs);
+                promise.completeInternal(RequestResult.ABORTED);
+                continue;
+            }
+
+            logger.trace("Running a background task check for {} with {}", cfs, cfs.getCompactionStrategy().getName());
+
+            CompletableFuture<Void> compactionTasks = startCompactionTasks(cfs);
+            if (compactionTasks == null)
+                compactionTasks = startUpgradeTasks(cfs);
+
+            if (compactionTasks != null)
+            {
+                compactionTasks.handle((ignored, throwable) -> {
+                    if (throwable != null)
+                    {
+                        logger.warn(String.format("Aborting compaction of %s due to error", cfs), throwable);
+                        handleCompactionError(throwable, cfs);
+                        promise.completeExceptionallyInternal(throwable);
+                    }
+                    else
+                    {
+                        logger.trace("Finished compaction for {}", cfs);
+                        promise.completeInternal(RequestResult.COMPLETED);
+                    }
+                    return null;
+                });
+
+                // The compaction strategy may return more subsequent tasks if we ask for them. Therefore, we request
+                // the compaction again on that CFS early (without waiting for the currently scheduled/started
+                // compaction tasks to finish). We can start them in the next check round if we have free slots
+                // in the compaction executor.
+                requestCompaction(cfs);
+            }
+            else
+            {
+                promise.completeInternal(RequestResult.NOT_NEEDED);
+            }
+        }
+    }
+
+    private boolean maybeScheduleNextCheck(Duration delay)
```

Comment:
Nit: we now always schedule with `Duration.ZERO`. Maybe fully revert fece018e4b9ac59330100c95b04c044702057b66?